File: frontend/lib/features/productivity/data/productivity_service.dart
import 'package:dio/dio.dart';

import '../../../core/network/dio_client.dart';
import '../../../models/productivity.dart';

class ProductivityService {
  ProductivityService(this._dioClient);

  final DioClient _dioClient;

  Dio get _dio => _dioClient.dio;

  Future<List<QuickNote>> listNotes() async {
    try {
      final response = await _dio.get('/productivity/notes');
      return (response.data as List<dynamic>).map((e) => QuickNote.fromJson(e as Map<String, dynamic>)).toList();
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<QuickNote> createNote(String content) async {
    try {
      final response = await _dio.post('/productivity/notes', data: {'content': content});
      return QuickNote.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<void> deleteNote(int id) async {
    try {
      await _dio.delete('/productivity/notes/$id');
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<List<TaskItem>> listTasks() async {
    try {
      final response = await _dio.get('/productivity/tasks');
      return (response.data as List<dynamic>).map((e) => TaskItem.fromJson(e as Map<String, dynamic>)).toList();
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<TaskItem> createTask({required String title, String priority = 'MEDIUM'}) async {
    try {
      final response = await _dio.post('/productivity/tasks', data: {'title': title, 'priority': priority});
      return TaskItem.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<TaskItem> updateTaskStatus(int id, String status) async {
    try {
      final response = await _dio.patch('/productivity/tasks/$id', data: {'status': status});
      return TaskItem.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }

  Future<void> deleteTask(int id) async {
    try {
      await _dio.delete('/productivity/tasks/$id');
    } on DioException catch (e) {
      throw DioClient.mapError(e);
    }
  }
}

File: backend/app/services/productivity_service.py
from __future__ import annotations

from app.core.exceptions import ForbiddenError, NotFoundError
from app.models.productivity import QuickNote, TaskItem
from app.repositories.productivity_repo import ProductivityRepository
from app.schemas.productivity import QuickNoteCreate, TaskItemCreate, TaskItemUpdate


class ProductivityService:
    def __init__(self, repo: ProductivityRepository) -> None:
        self.repo = repo

    async def list_notes(self, user_id: int) -> list[QuickNote]:
        return await self.repo.list_notes(user_id)

    async def create_note(self, user_id: int, data: QuickNoteCreate) -> QuickNote:
        note = QuickNote(user_id=user_id, **data.model_dump())
        return await self.repo.create_note(note)

    async def update_note(self, note_id: int, user_id: int, content: str) -> QuickNote:
        note = await self.repo.get_note(note_id)
        if not note:
            raise NotFoundError("Note not found")
        if note.user_id != user_id:
            raise ForbiddenError("You do not own this note")
        note.content = content
        return await self.repo.update_note(note)

    async def delete_note(self, note_id: int, user_id: int) -> None:
        note = await self.repo.get_note(note_id)
        if not note:
            raise NotFoundError("Note not found")
        if note.user_id != user_id:
            raise ForbiddenError("You do not own this note")
        await self.repo.delete_note(note)

    async def list_tasks(self, user_id: int) -> list[TaskItem]:
        return await self.repo.list_tasks(user_id)

    async def create_task(self, user_id: int, data: TaskItemCreate) -> TaskItem:
        task = TaskItem(user_id=user_id, **data.model_dump())
        return await self.repo.create_task(task)

    async def update_task(self, task_id: int, user_id: int, data: TaskItemUpdate) -> TaskItem:
        task = await self.repo.get_task(task_id)
        if not task:
            raise NotFoundError("Task not found")
        if task.user_id != user_id:
            raise ForbiddenError("You do not own this task")
        for key, value in data.model_dump(exclude_unset=True).items():
            setattr(task, key, value)
        return await self.repo.update_task(task)

    async def delete_task(self, task_id: int, user_id: int) -> None:
        task = await self.repo.get_task(task_id)
        if not task:
            raise NotFoundError("Task not found")
        if task.user_id != user_id:
            raise ForbiddenError("You do not own this task")
        await self.repo.delete_task(task)
